### PR TITLE
Fix allowedAction comparison and root enforcement

### LINKED ISSUE #89: `hasValidAllowedAction` rejects an array child against a string parent, even for an identical action set
`allowedAction` is explicitly allowed to be either a string or a non-empty array (`utils.js:563-569`). `hasValidAllowedAction` handles three of the four resulting type combinations, but not a string parent with an array child, so a delegation that narrows nothing is rejected as "less restrictive than its parent".

```js
// utils.js:492-509
export function hasValidAllowedAction({allowedAction, parentAllowedAction}) {
  if(!parentAllowedAction) {
    return true;
  }
  if(Array.isArray(parentAllowedAction)) {
    if(Array.isArray(allowedAction)) {
      return allowedAction.every(a => parentAllowedAction.includes(a));
    }
    return parentAllowedAction.includes(allowedAction);
  }
  // require exact match
  return (parentAllowedAction === allowedAction);
}
```

| parent | child | result |
|---|---|---|
| `['write']` | `['write']` | passes |
| `['write']` | `'write'` | passes |
| `'write'` | `'write'` | passes |
| `'write'` | `['write']` | **rejected** |

All four express the same authority, and the fourth is the one `@digitalbazaar/ezcap` produces by default.

### How it is reached

`ZcapClient.delegate()` inherits the parent's `allowedAction` when the caller does not pass `allowedActions`, then coerces a string to an array before building the child:

```js
// ezcap 4.3.0, ZcapClient.js:214-226, abridged
if(allowedActions === undefined) {
  allowedActions = capability.allowedAction || [];   // 'write'
}
if(typeof allowedActions === 'string') {
  allowedActions = [allowedActions];                 // ['write']
}
```

So delegating a capability whose `allowedAction` is a bare string, without asking to change the actions at all, throws. The check also runs at mint time in `CapabilityDelegation.update()` (`:157-164`), so the failure happens during `delegate()` rather than later at verification.

### Reproduction

zcap 9.0.1, ezcap 4.3.0, node 24. Parent is a root capability whose only unusual property is that `allowedAction` is a string:

```js
const root = {
  '@context': 'https://w3id.org/zcap/v1',
  id: `urn:zcap:root:${encodeURIComponent(target)}`,
  controller,
  invocationTarget: target,
  allowedAction: 'write'      // a string, per utils.js:563-569
};

await client.delegate({
  capability: root,
  controller: delegateDid,
  expires: '2027-01-01T00:00:00Z'
  // allowedActions omitted: "do not change the actions"
});
```

Result:

```
parent allowedAction: "write"
THREW: The "allowedAction" in a delegated capability must not be less restrictive than its parent.
```

Changing only the parent's `allowedAction` to `['write']`, with the delegation call untouched, succeeds and yields a child with `allowedAction: ['write']`.

### Suggested fix

Normalize both sides before comparing, so the check is about the action set rather than the JSON type:

```js
export function hasValidAllowedAction({allowedAction, parentAllowedAction}) {
  if(!parentAllowedAction) {
    return true;
  }
  const parent = Array.isArray(parentAllowedAction) ?
    parentAllowedAction : [parentAllowedAction];
  const child = Array.isArray(allowedAction) ?
    allowedAction : [allowedAction];
  return child.every(a => parent.includes(a));
}
```

This preserves the behavior of all three currently-passing combinations.

### Notes

Filing here rather than on `ezcap` because the asymmetry is in this comparison, and normalizing it fixes the case for any caller, not only ones going through `ZcapClient`. Happy to open a PR if the approach looks right. If the string form is instead considered deprecated in favor of arrays, rejecting a string `allowedAction` in `checkCapability` would at least make that explicit, since the current behavior accepts the document and then refuses to delegate from it.

### PR DESCRIPTION
**Pull request by jameseaster**
Compares `allowedAction` sets rather than JSON types so a delegation that narrows nothing succeeds whether each side spells its actions as a string or an array, and seeds `parentAllowedAction` from the root zcap so a root's `allowedAction` is enforced by the verifier and not only when a delegation proof is generated.

Fixes #89

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,19 @@
 # @digitalbazaar/zcap ChangeLog
 
+## 10.0.0 - 2026-09-dd
+
+### Changed
+- **BREAKING**: Enforce a root capability's `allowedAction` when verifying a
```
Comment (dlongley) [inline CHANGELOG.md]: I think this should be a patch. A delegated zcap that was invalid that this library didn't catch (even if properly signed by the delegator) is still invalid and this patch will now properly reject it. Those zcaps shouldn't have been accepted to begin with, so this is a fix.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (jameseaster) [reply CHANGELOG.md]: Updated changelog here: 98fdac6

File: lib/utils.js [hunk 2]
```diff
@@ -496,16 +496,20 @@ export function hasValidAllowedAction({allowedAction, parentAllowedAction}) {
...
-  return (parentAllowedAction === allowedAction);
+  /* Note: `allowedAction` can be expressed as either a string or a non-empty
+  array (see `checkCapability`), so compare the action sets, not the JSON
+  types used to express them. */
+  const parentActions = getAllowedActions(
+    {capability: {allowedAction: parentAllowedAction}});
+  const actions = getAllowedActions({capability: {allowedAction}});
+
+  // parent's `allowedAction` must include every one from child's
+  return actions.every(a => parentActions.includes(a));
```
Comment (dlongley) [inline lib/utils.js]: Can we add a test that covers an explicitly provided, but empty array for `allowedAction`? Added tests here should at least include negative tests where the child uses an empty array but the parent used a string or a non-empty array. It seems like an empty array in the child could pass here when it shouldn't.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/utils.js (hunk 2 above)
Comment (jameseaster) [reply lib/utils.js]: Agreed, that the helper itself would return `true` there, but an empty array can't reach it: `checkCapability` rejects any `allowedAction` that isn't a string or non-empty array ([utils.js#L578](https://github.com/digitalbazaar/zcap/blob/105adc4/lib/utils.js#L578)) and runs first on both paths - before `hasValidAllowedAction` in `CapabilityDelegation.update` ([L145](https://github.com/digitalbazaar/zcap/blob/105adc4/lib/CapabilityDelegation.js#L145) vs [L160](https://github.com/digitalbazaar/zcap/blob/105adc4/lib/CapabilityDelegation.js#L160)), and on every zcap in the dereferenced chain ([utils.js#L339](https://github.com/digitalbazaar/zcap/blob/105adc4/lib/utils.js#L339), [#L412](https://github.com/digitalbazaar/zcap/blob/105adc4/lib/utils.js#L412)) before `_verifyCapabilityChain` ever compares actions - which the three tests in 105adc4 now pin down at both layers.

**Review by dlongley — APPROVED** [review]
